### PR TITLE
Added dummy routetable for network policy only mode

### LINKED ISSUE #5247: Add configuration option to disable route programming from calico/node
Hi everyone! This is not a bug, but more like a feature request of sorts.

## tl;dr

We want to propose adding a configuration to disable any interaction between `calico-node` (through `felix`) and the routing tables. This is to allow running Calico as a NetworkPolicy controller only.

## Longer version

We want to know how hard would it be to modify `felix` to allow running `calico-node` in a true Network Policy only mode. That is, to have `calico-node` only interact with `iptables` and `ipset` **without ever modifying any routing table**.

We managed to have it running that way by replacing [this assignment](https://github.com/projectcalico/felix/blob/092e3b76f40c0eba36c94a78752abc2856878509/dataplane/linux/int_dataplane.go#L676) with a dummy interface (only here, because we needed it for ipv4 and not ipv6). That prevented `calico-node` to touching the routing table, while letting it configure `iptables` normally.

As a side note, the names of the interfaces still have to match what Calico expects (both the prefix, and the hashed Pod name/namespace)

Of course, that was only a quick PoC; but a proper change could be considered by adding a configuration parameter to disable the route table sync (maybe `EnableRouteTableManager=false`). The purpose of this issue is to discuss:

0) Is there already a better way of running `calico-node` as Network Policy only mode that we've may have overlooked?
1) Do you think it would be worthwhile to have such configuration to disable all interaction with the routing table?
2) Do you see any issues with this approach?
3) Would it suffice to track down usages of `routeTableSyncer` and `routeTable` interfaces and replace them with noop ones? There are not many of these, and disabling should be straightforward.

We believe that this could open the door to more users that only want/need a NetworkPolicy manager; since there are not many out of the box solutions for *only that*.

If you think it's viable, I'd love to work on a PR and send it your way!

Let us know any thoughts you may have on this, all feedback is appreciated.
I'll leave more context below, fwiw.

## More context

We are running a k8s cluster with a homemade CNI plugin (the details of that are not too relevant, but we have some specific requirement that require us to have multiple IPs per Pod). This plugin runs in a similar fashion to the `ptp` standard plugin: it creates routes in the host's routing table and expects someone else to propagate the routes across the cluster. We are using `frr` for that.

Up to recently we were using `kube-router` as an implementation for NetworkPolicies, but it turned out to not scale well with the total number of Policies. So we looked for alternatives and one of them was Calico.

We wanted to run **Calico as a NetworkPolicy controller only**, and found out that there was no easy way of doing so.

We tried [calico-policy-only.yaml](https://docs.projectcalico.org/v3.17/manifests/calico-policy-only.yaml), and deleted `calico-kube-controllers` from there. Up to that point we had one `calico-node` DaemonSet, and one `calico-typha` Deployment; and with that NetworkPolicies were implemented by Calico. (The folks at cybozu-go had [a similar situation](https://github.com/cybozu-go/neco-apps/blob/main/docs/maintenance.md#network-policy-calico), we use it as a base)

However, that setup also messed up with our routes, particularly it deleted any route we had for those "extra IPs" (i.e. any IP we assign to the Pod via CNI, that does not reflect through the k8s apiserver). In particular, we noticed that the `endpoint manager` was active and attempting a sync of the routing table with every pod created/destroyed. This is of course desirable if Calico is used for both routing and NetworkPolicies, but we already had our solution for routing that we sadly cannot change. So we tried looking for a way to disable that behaviour.

This is an example of the logs of `calico-node` deleting routes. Something to notice here is that we changed the interface names of all our veth to match what Calico expects. This lead us to having Calico **only delete the extra IPs routes**. Any route that matched with the Pod IP will be validated by Calico and left intact.

```
2021-01-25 20:24:33.019 [INFO][64] felix/route_table.go 865: Remove old route dest=10.228.200.32/32 ifaceName="veth74985cde457" ifaceRegex="^veth.*" ipVersion=0x4 routeProblems=[]string{"unexpected route"}
2021-01-25 20:24:33.019 [INFO][64] felix/conntrack.go 90: Removing conntrack flows ip=10.228.200.32
```

We didn't find a way of disabling this behavior from the configuration; so we took a look at the code to see how hard would it be to run `calico-node` but with the endpoint manager disabled. It looks like that the code that the manager that controls the routing table is initialized [here](https://github.com/projectcalico/felix/blob/092e3b76f40c0eba36c94a78752abc2856878509/dataplane/linux/int_dataplane.go#L676). We attempted replacing the current implementation `routetable` with a dummy one that did nothing and that seemed to solve our issues.

So the origin of this issue was to see if it would be worthwhile to work this into a proper configuration, if anyone else would be interested in that.

### PR DESCRIPTION
**Pull request by juanfresia**
## Description

This PR adds a config flag for `felix` named `RouteSyncDisabled`:
* The default value is `false`.
* Setting this to `true` will prevent `felix` from doing any updates on the route table.
* This is applied to all instances of `RouteTable` (IPv4, IPv6, VXLAN and Wireguard).

In order to make sure no operations are done against the linux routing table, a new implementation of `RouteTable` called `DummyTable` was added. While using the `DummyTable` all operations are skipped, enabling the use of `felix` as a network policy controller only.

## Related issues/PRs

More information on the context for this change can be found in #5247.

## Release Note

```release-note
Added dummy routetable for network policy only mode
```

### CODE REVIEW

**Comment by song-jiang** [comment]
Thanks @juanfresia for the PR. There are multiple places need to be updated in order to add a Felix configuration. 
- ./api directory. https://github.com/projectcalico/calico/blob/master/api/pkg/apis/projectcalico/v3/felixconfig.go and auto generated files
- ./libcalico directory.  CI and auto generated files

File: felix/dataplane/linux/int_dataplane.go [hunk 1]
```diff
@@ -709,9 +710,17 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 		}
 	}
 
-	routeTableV4 := routetable.New(interfaceRegexes, 4, false, config.NetlinkTimeout,
-		config.DeviceRouteSourceAddress, config.DeviceRouteProtocol, config.RemoveExternalRoutes, 0,
-		dp.loopSummarizer)
+	var routeTableV4 routeTable
+	routeTableV4 = &routetable.DummyTable{}
```
Comment (fasaxc) [inline felix/dataplane/linux/int_dataplane.go:714]: This only covers one instance of the RouteTable, there can be multiple:
* One for IPv4, one for IPv6
* One for VXLAN
* etc.

File: felix/config/config_params.go [hunk 2]
```diff
@@ -330,7 +330,8 @@ type Config struct {
 	// - calicoIPAM: use IPAM data to contruct routes.
 	RouteSource string `config:"oneof(WorkloadIPs,CalicoIPAM);CalicoIPAM"`
 
-	RouteTableRange idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	RouteTableRange  idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	RouteSyncEnabled bool               `config:"bool;true"`
```
Comment (song-jiang) [inline felix/config/config_params.go:334]: I would suggest to use `RouteSyncDisabled` with default value `false`. The configuration item is part of FelixConfiguration resource and should be omit-empty.

**Review by song-jiang — CHANGES_REQUESTED** [review]

**Comment by juanfresia** [comment]
Thanks you both for the comments, I'll send a commit to address them shortly

**Comment by juanfresia** [comment]
Hi @song-jiang @fasaxc sorry for the delay. I've just updated the PR with the requested changes.
* I added the same change for `IPv6, VXLAN and Wireguard, which are all the uses of RouteTable I could find.
  * I think IPv6 and IPv4 are quite similar; but I'm not super familiar with the other two. If you know of any side effects of this change for those, you can point me in the right direction to fix them; or maybe it'd be an alternative to keep this flag from affecting VXLAN and Wireguard.
* I also changed the semantics to be `RouteSyncDisabled` with a `omitempty` config and a `false` default.
* `api` and `libcalico-go` sub-projects have been updated (hopefully, in the right way).

Also updated the PR's original description.

Let me know if there are any more needed changes/or ideas to be added here. Since this is basically adding a Dummy implementation of an interface, I'm not quite sure which tests would be valuable here; let me know if you have any ideas here as well.

File: felix/dataplane/linux/int_dataplane.go [hunk 3]
```diff
@@ -434,9 +435,16 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 	dp.ipSets = append(dp.ipSets, ipSetsV4)
 
 	if config.RulesConfig.VXLANEnabled {
-		routeTableVXLAN := routetable.New([]string{"^vxlan.calico$"}, 4, true, config.NetlinkTimeout,
-			config.DeviceRouteSourceAddress, config.DeviceRouteProtocol, true, 0,
-			dp.loopSummarizer)
+		var routeTableVXLAN routeTable
+		if !config.RouteSyncDisabled {
+			log.Info("RouteSyncDisabled is false.")
```
Comment (song-jiang) [inline felix/dataplane/linux/int_dataplane.go:440]: In most cases `RouteSyncDisabled` will be false so I suggest to use `log.Debug` . Similar suggestions apply to other RouteTable below.

File: felix/dataplane/linux/vxlan_mgr.go [hunk 4]
```diff
@@ -105,31 +105,47 @@ func newVXLANManager(
...
+			false,
+			0,
+			opRecorder,
+		)
+	} else {
+		log.Info("RouteSyncDisabled is true, using DummyTable.")
+		brt = &routetable.DummyTable{}
+	}
+
+	noEncapRTConstruct := func(interfaceRegexes []string, ipVersion uint8, vxlan bool, netlinkTimeout time.Duration,
```
Comment (song-jiang) [inline felix/dataplane/linux/vxlan_mgr.go:127]: Maybe move setting of `noEncapRTConstruct` to `if !dpConfig.RouteSyncDisabled` above?

File: felix/wireguard/route_table.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+// routeTable is the interface provided by the standard routetable module used to progam the RIB.
+type routeTable interface {
+	routeTableSyncer
+	SetRoutes(ifaceName string, targets []routetable.Target)
+	SetL2Routes(ifaceName string, targets []routetable.L2Target)
+	RouteRemove(ifaceName string, cidr ip.CIDR)
+	RouteUpdate(ifaceName string, target routetable.Target)
+}
```
Comment (song-jiang) [inline felix/wireguard/route_table.go:24]: I'm not sure why you need to add this file.

File: felix/wireguard/wireguard.go [hunk 6]
```diff
@@ -150,7 +150,7 @@ type Wireguard struct {
 	cidrToNodeName map[ip.CIDR]string
 
 	// Wireguard routing table and rule managers
-	routetable *routetable.RouteTable
+	routetable routeTable
```
Comment (song-jiang) [inline felix/wireguard/wireguard.go:153]: Not sure about the change here.

**Comment by song-jiang** [comment]
@juanfresia I've reviewed your changes and added couple of comments. Thanks!

File: felix/dataplane/linux/int_dataplane.go (hunk 3 above)
Comment (juanfresia) [reply felix/dataplane/linux/int_dataplane.go:440]: Sounds good, I'll change it for a `log.Debug` when we hit the default value for `RouteSyncDisabled`

File: felix/wireguard/route_table.go (hunk 5 above)
Comment (juanfresia) [reply felix/wireguard/route_table.go:24]: So, Wireguard was using the `RouteTable` struct directly from the `routetable` package, and that meant that I could not override it directly with the `DummyTable`. All other uses of `RouteTable` went through the `routeTable` interface, from `intdataplane` package.

That interface did not include the `RouteRemove` and `RouteUpdate` methods `Wireguard` uses; and I didn't want to introduce a dependency between `wireguard` and `intdataplane` packages for this; so I replicated the interface with just the methods `wireguard` needs, checking that both `routetable.RouteTable` and `DummyTable` implements them.

That was my logic, but of course I could be missing something, do you have any suggestions on how would an alternative look like?

File: felix/wireguard/wireguard.go (hunk 6 above)
Comment (juanfresia) [reply felix/wireguard/wireguard.go:153]: See the other comment, changed `routetable.RouteTable` struct with an `wireguard.routeTable` interface.

File: api/pkg/apis/projectcalico/v3/felixconfig.go [hunk 7]
```diff
@@ -370,6 +370,10 @@ type FelixConfigurationSpec struct {
 	// specifies the indices of the route tables that Calico should use.
 	RouteTableRange *RouteTableRange `json:"routeTableRange,omitempty" validate:"omitempty"`
 
+	// RouteSyncDisabled will disable all operations performed on the route table. Set to true to
+	// run in network-policy mode only.
+	RouteSyncDisabled *bool `json:"routeSuncDisabled,omitempty"`
```
Comment (M00nF1sh) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:375]: typo: routeSuncDisabled-> routeSyncDisabled

File: libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml [hunk 8]
```diff
@@ -496,6 +496,10 @@ spec:
                   information. - WorkloadIPs: use workload endpoints to construct
                   routes. - CalicoIPAM: the default - use IPAM data to construct routes.'
                 type: string
+              routeSuncDisabled:
```
Comment (M00nF1sh) [inline libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml:499]: typo: routeSuncDisabled-> routeSyncDisabled

Comment (juanfresia) [reply libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml]: Thanks! Fixed now.

**Comment by juanfresia** [comment]
@song-jiang Just fixed the log level as per your comment. I had left you some question regarding your other comments, what do you think would be the best alternative?

**Comment by song-jiang** [comment]
@juanfresia I'm sorry I somehow missed this thread and have been on vacation lately. Thanks for all your work. 
I'd like to get input from @mikestephen on changes related to wireguard, but the rest code changes LGTM. 

By the way, could you please rebase on master branch again?

**Comment by juanfresia** [comment]
Thanks @song-jiang , I'll rebase the changes

**Comment by caseydavenport** [comment]
@juanfresia looks like this still needs a rebase.

**Comment by juanfresia** [comment]
Hi @mikestephen! I believe @song-jiang wanted your input on this change, do you have any thoughs on this?

**Comment by song-jiang** [comment]
@juanfresia I had a chat with @mikestephen and @caseydavenport , we think at this point, the PR is good to get merged. I'll raise a follow up PR to address two improvements:
1. Fix Calico doc for the new

**Comment by song-jiang** [comment]
@juanfresia I had a chat with @mikestephen and @caseydavenport , we think at this point, the PR is good to get merged. I'll raise a follow up PR to address two improvements:
1. Fix Calico doc for the new

**Review by song-jiang — APPROVED** [review]

**Comment by mgleung** [comment]
@juanfresia looks like this needs another rebase and then it's ready to go with the passing tests.

**Comment by juanfresia** [comment]
Thanks @mgleung, I'll sync it again against `master` shortly

**Comment by juanfresia** [comment]
Hi @song-jiang, you mentioned some documentation PRs missing; is there an example I can use as an example? Do you think we can merge this PR or should we keep it open until documentation is ready?

Other than that, is there anything we are missing to merge this change?

**Comment by song-jiang** [comment]
@juanfresia Sorry I missed your comment. I think we can just merge this PR. We need to do the following:
1. Fix Calico doc for the new felix parameter.
2. Make wireguard start using the routeTable interface. 

I can take care both point but please let me know if you are happy to do the first one. Thanks.

**Pull request merged**